### ISSUE TITLE
Discrepancy of max_size between string and bytes types

### ISSUE THREAD

**Issue by GoogleCodeExporter** [nanopb/nanopb#107]
```
Suppose I have the following:

foo.proto:

message Foo {
    required bytes b = 1;
    required string s = 2;
}

foo.options:

Foo.b max_size:10
Foo.s max_size:10

Nanopb will generate the following:

typedef struct {
    size_t size;
    uint8_t bytes[10];
} Foo_b_t;

typedef struct _Foo {
    Foo_b_t b;
    char s[10];
} Foo;

While this might seem fine for a C programmer, it doesn't feel logical for a 
protocol designer. The issue is that one is able to store 10 bytes of data in 
"b", but only 9 characters in "s" (because of the zero-terminated string). 

My suggestion would be to generate

    char s[10+1];

for the mentioned case, so even the string can hold 10 characters.

I think the suggestion follows the least surprise principle. 
```

Original issue reported on code.google.com by `rusn...@gmail.com` on 2 Mar 2014 at 5:19

**Comment by jnread** [comment]
This seems to have dropped off the radar a little, but it's still biting me.  I use nanopb to talk to C# code, so don't really want to enforce zero terminators in the protocol definition itself.

The "+1" method would work at the expense of alignment as you say, but it would also be nice is nanopb automatically added the terminator.

Alternatively, strings could be treated in much the same way as you treat byte arrays now, with an explicit length and leaving the user to ensure the string is terminated correctly when used.

**Comment by PetteriAimonen** [comment]
Hmm, the C# code shouldn't care about the nanopb max_size options anyway. You can put them in a separate options file also, if you don't want to have them in the .proto.

**Comment by jnread** [comment]
Yes, sorry perhaps I should have been more clear.  We may agree on a max string length that the C# code adheres to, but it doesn't currently correspond to the max size in the .proto.  We effectively use the .proto as a contract between the C and C# developers, to provide a single source of truth.

eg. we agree to only accept a string of max length 10, but the .proto has to have a max_size of 11, which leads to confusion as the contract isn't completely clear. 

It would perhaps be an alternative idea to have a max_length option for strings, which differs from max_size in that the generated C code has additional space for a terminating zero - but the "length" is then usable contractually?  This would avoid any issues with backwards compatibility for existing nanopb users.

I don't wish to make a mountain out of a molehill here, but I would gladly look at implementing this myself if it's considered worthwhile.

**Comment by PetteriAimonen** [comment]
Ah, yeah, makes sense.

Adding a `max_length` option is actually a great idea and easy to implement. Should be fixed now :)

**Comment by jnread** [comment]
Awesome!  Thank you @PetteriAimonen !

**Comment by PetteriAimonen** [comment]
Fix released in nanopb-0.3.8.
